kernel/src/engine/arrow_utils.rs

```diff
@@ -66,17 +69,63 @@ pub(crate) fn make_arrow_error(s: impl Into<String>) -> Error {
     .with_backtrace()
 }
 
+/// Prepares to enumerate row indexes of rows in a parquet file, accounting for row group skipping.
+pub(crate) struct RowIndexBuilder {
+    row_group_starting_row_offsets: Vec<Range<i64>>,
```

Comment:
nit: I think `row_group_row_indexes`, or `row_group_row_index_ranges` or similar, would make more sense as a name. This isn't really the offsets, it's the ranges of rows that each group covers.